### ISSUE TITLE
The index route on the RoutablePageMixin will be run if your routable page model is extended and your 'index' route is defined only on the parent model

### ISSUE THREAD

**Issue by douglasd** [wagtail/wagtail#3860]
### Issue Summary

If you define an abstract Page model and decorate a function with
@route(r'^$')
and then extend that model into other RoutableMixin pages but don't define the @route(r'^$') on the extending pages then the abstract model route is ignored and the new RoutablePageMixin index route is run.

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. Edit models.py as follows:
- add an abstract page model and decorate a function with the @route(r'^$') route
- define a new routable page model that extends the abstract model but does not define its own @route(r'^$') handler
- add something specific to the abstract class's handler function like a custom template or context details
3. Save and view the extended page.
4. The abstract model's function will not run but the mixin 'index' route will.

Our use case is that we have a pair of routable pages that extend an abstract page. We didn't see the need to duplicate the route handler just to add some specific context to the page render so we defined identical functions on the extending page classes and call those from the 'shared' abstract handler function.

eg:

```
@route(r'^$')
 def page(self, request, *args, **kwargs):
        ctx = {'breadcrumbs': self._breadcrumbs()}
        ctx.update(self.get_context(request, *args, **kwargs))
        self.featured_deals_params = self._featured_deals_params(request, ship_id = self.ship_id)
        return TemplateResponse(
            request,
            self.page_template,
            ctx
        )
```

In our case, featured_deals_params is a function that is on all extending classes but returns details specific to the extending classes execution of it.

I realize that this is understood behaviour of inheritance / mixin but I wasn't sure if this was intended. We were hit by this in upgrading from 1.4 to 1.6. We can move the functions down to the extending classes but I thought it best to bring it up.

**Comment by gasman** [comment]
Hi @douglasd - thanks for the report.

Please can you give an example of a complete models.py file that demonstrates this bug? I tried with the one below following your instructions, but it's working as expected - creating and viewing an instance of ConcreteRoutablePage gives the response "This is the abstract home page".

    from __future__ import absolute_import, unicode_literals

    from django.db import models
    from django.http import HttpResponse

    from wagtail.wagtailcore.models import Page
    from wagtail.contrib.wagtailroutablepage.models import RoutablePageMixin, route


    class HomePage(Page):
        pass


    class AbstractRoutablePage(RoutablePageMixin, Page):
        @route(r'^$')
        def home(self, request):
            return HttpResponse("This is the abstract home page")

        class Meta:
            abstract = True


    class ConcreteRoutablePage(AbstractRoutablePage):
        pass

**Comment by douglasd** [comment]
Hi @gasman, I see what it is now. We aren't mixing in the RoutablePageMixin on the abstract class, we're mixing it into the extending class. If you move the mixin from your AbstractRoutablePage to your ConcreteRoutablePage you'll see the behaviour I'm talking about.

I can't recall exactly why we mixed it into the extended class but I do seem to recall that since we use another mixin and the inheritance chain of the page is longer than just a single abstract class and an extending concrete class (We have Page -> BasePage (some all page level items) -> AbstractRoutablePage -> ConcretePage mixing in RoutableMixin and another one) that we needed to move the mixins around a bit to make them work the way we wanted.

Ultimately, for better or worse, we ended up with a setup where the abstract page with the shared @route(r'^$') handler wasn't the class mixing in RoutablePage.

I will mess with our setup and see if moving the mixin to the abstract class will work because that seems like the right solution. The documentation says to use the mixin, the class using it should inherit from wagtail.wagtailcore.models.Page and we're we to be doing that we'd be fine.
